NSBrute.py

```diff
@@ -88,45 +96,107 @@ def myPrint(text, type):
+				created_zones.remove(hosted_zone_id)
+				print ""
+				break
+		except Exception as e:
+			myPrint("Exceptional behaviour observed while creating the zone.", "ERROR")
+			myPrint("Trying Again!","ERROR")
+			if new_zone != 0:
+				new_zone.delete()
 			continue
-		nsAWS=new_zone.nameservers
-		myPrint("Created a new zone with following NS: ","INFO_WS")
-		myPrint("".join(nsAWS),"INFO_WS")
-		intersection=set(nsAWS).intersection(set(targetNS))
-		if(len(intersection)==0):
-			myPrint("No common NS found, deleting new zone","ERROR")
-			print ""
-			new_zone.delete()
-		else:
-			myPrint("Successful attempt after "+str(counter)+" iterations.","SECURE")
-			myPrint("Check your AWS account, the work is done!","SECURE")
-			print ""
-			exit(0)
-	except Exception as e:
-		myPrint("Exceptional behaviour observed while creating the zone.", "ERROR")
-		myPrint("Trying Again!","ERROR")
-		if new_zone != 0:
-			new_zone.delete()
-		continue
 
+except KeyboardInterrupt:
+	quit = raw_input("Would you like to quit without having deleted all hosted zones? Put 'y' if you want to quit now without having deleted all hosted zones. Put any other sequence of characters to quit immediately: ")
+	if (len(created_zones) != 0 and quit == "y") or forceDelete:
+		command = "AWS_ACCESS_KEY_ID="+accessKey+" AWS_SECRET_ACCESS_KEY="+secretKey+" aws route53 list-hosted-zones"
+		out = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
+		stdout,stderr = out.communicate()
+		json_data = None
+
+		if stdout != 'false':
+			json_data = json.loads(stdout)
+		
+		zones_to_be_removed = []
+		zones_for_account = []
+		
+		for zone in json_data["HostedZones"]:
+			zones_for_account.append(str(zone["Id"].replace("/hostedzone/","")))
+
+		if len(successful_zone) != 0:
+			if successful_zone[0] in created_zones:
+				created_zones.remove(successful_zone[0])
+		
+		for zone in created_zones:
+			if zone in zones_for_account:
+				zones_to_be_removed.append(zone)
+		
+		for zone in zones_to_be_removed:
+			command = "AWS_ACCESS_KEY_ID="+accessKey+" AWS_SECRET_ACCESS_KEY="+secretKey+" aws route53 delete-hosted-zone --id " + str(zone)
+			out = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
+			stdout,stderr = out.communicate()
+
+	else:
+		exit()
+
+command = "AWS_ACCESS_KEY_ID="+accessKey+" AWS_SECRET_ACCESS_KEY="+secretKey+" aws route53 list-hosted-zones"
+out = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
+stdout,stderr = out.communicate()
+json_data = None
+
+if stdout != 'false':
+	json_data = json.loads(stdout)
+
+zones_to_be_removed = []
+zones_for_account = []
+
+for zone in json_data["HostedZones"]:
+	zones_for_account.append(str(zone["Id"].replace("/hostedzone/","")))
+
+if len(successful_zone) != 0:
+	if successful_zone[0] in created_zones:
+		created_zones.remove(successful_zone[0])
+
+for zone in created_zones:
+	if zone in zones_for_account:
```

Comment:
I guess this would always return true. I didn't understand when we already have the list of created zones why do we need to fetch all the hosted zones. Can't we simply delete all the created zones -successful_zone?

Reply:
Good question, at first thought I thought the same. The trouble is when the tool successfully deletes the zone not using the direct `subprocess.Popen` call that I am making, then when the `subprocess.Popen` call runs later, it can sometimes cause the script to prematurely halt due to lack of being able to handle an exception. As such, I put this logic in and it seems to function properly every time.